### ISSUE TITLE
Nightly Builds for Windows and MacOS

### ISSUE THREAD

**Issue by LinuxDonald** [fifengine/fifengine#888]
What about doing Nightly oder git commit based builds for Windows and MacOS because its not easy to build both (fife and fifechan) on this platforms.

@stubb what do you think? (ps please look at your spam folder again ;) )

**Comment by stubb** [comment]
We dont have such build slaves yet. For os x its easy to build fife and fifechan, not sure about creating an app out of it. Never build a windows installer yet, would take some time. But afterwards should be easily run automatic.

**Comment by jakoch** [comment]
Hey!

> What about doing Nightly oder git commit based builds for Windows and MacOS because its not easy to build both (fife and fifechan) on this platforms.

**Nightly Builds**
Nightly builds are periodical builds triggered by a time event, often a cronjob. This feature is really new on Travis and wasn't available on until May and is still only available on-request(!). Referencing: https://docs.travis-ci.com/user/cron-jobs/

Windows is still not supported by Travis. So, for Windows builds the alternative would be to setup Appveyor-CI based builds and enable scheduled building to get Nightlies. Referencing: http://www.appveyor.com/docs/build-configuration#scheduled-builds

**Git commit based builds**
The project already has "Git commit based builds", but these are not deployed. 
We would simply need to enable automatic deployments on Travis (and Appveyor) to Github Releases on every commit. This needs a change in the versioning scheme, e.g. `v1.2.3-hj43234` (version and git commit hash). 

The follow-up question is then, do you want to enable deployment from every branch or only from specific ones, e.g. only master?. I would suggest to only enable "master".

**To sum-up**: Its possible to build Nightly's for Linux and Mac on Travis and for Windows on Appveyor. 
The two main things needed in the build configuration are (a) automatic deployment and (b) scheduled building.

Sidenote:  I'm not a fan of time-based nor git-commit/rolling builds on projects with low commit and change frequency, because of "spam" releases. You would increase the release frequency by turning on nightly builds, but that wouldn't change the change frequency. Given the low change frequency, it would be better to let features and changes pile up and then do for instance a release with a larger interval, e.g. a `Monthly` release, instead of a daily `Nightly` one without any changes (probably). 

----

**Windows Installers**

There are two installers here: https://github.com/fifengine/fifengine/tree/master/build/win32
One is for Fife and the other is for the Development SDK.
But they are really out of date and not build automatically.
Also the manually build installers and files are still hosted on SourceForge: https://sourceforge.net/projects/fife/files/active/

> Never build a windows installer yet, would take some time.

I would like to help to create automated installer releases.
This enables the project to release Fife together with the DevSDK installer for a tagged release.

After a successful build/compilation of FifeEngine on Appveyor, 
we can build the python module installer and 
then build the windows installer as a second step and release it to Github Releases.

It would be nice, if we could go over all steps of the Windows builds again to clean things out, before we start to work on build and deployment automation.

Best, Jens

**Comment by prock-fife** [comment]
@jakoch thanks for the detailed writeup!  I agree with your approach with using travis builds for Linux and OSX, and Appveyor for WIndows.  We have some done some work recently which should help simplify the packaging process as I would only expect we package FIFE itself and not include the demos (including their resources) and/or the editor.  In the past we would package everything so the install builder will need to be updated and I'm not against replacing it completely with something else..  Unfortunately the dev kit is way out of date so that needs to be refreshed and somehow maintained.  We are working on moving from scons to cmake as well which might simplify the build process.  I would suggest you work with cmake branch to start as that is the direction we are moving.

As far as the build frequency I agree that fewer is better in our case because of the number of changes.  I would even venture to say quarterly would be sufficient.

On a side note I would also agree that we should move to GitHub releases and stop using sourceforge.

Some additional reference from the archive wiki for release management:

http://archivewiki.fifengine.net/Release_Management
http://archivewiki.fifengine.net/Release_packaging

If you would like to help we will be more than happy to accept your offer.  If you require anything at all from the team please let us know.  I'd imagine we'll need to set up some access for you at some point to help make the process easier for you.

**Comment by jakoch** [comment]
Hey Prock!

Thank you for updating me on the latest changes.

> ... simplify the packaging process as I would only expect we package FIFE itself and not include the demos (including their resources) and/or the editor.

This enables users to download the things they really need and it cuts package size down.
Good move.

> moving from scons to cmake

Very good decision.

I would suggest to push the cmake related changes, while keeping Scons in the background for safety and then simply do a 0.4.0 release using the new cmake based build system. 

> I would even venture to say quarterly would be sufficient.

That's better.

> If you would like to help we will be more than happy to accept your offer. If you require anything at all from the team please let us know. I'd imagine we'll need to set up some access for you at some point to help make the process easier for you.

I'll work in a personal fork to punch the Appveyor stuff out and then send a squashed PR.
Let's see how far i get...

--

Sidenote/off-topic: **Adding milestones to issues** 
You have marked [a lot of tickets belonging to the 0.4.0 milestone](https://github.com/fifengine/fifengine/issues?milestone=10). When closing all the issues is a must (to be done with that milestone), you will probably never get the 0.4.0 release out. If it's not a must, then planning failed once more and you end up re-scheduling all the tickets to the next milestone and then probably again... It's probably the dream of every open source project that things get done this way, but it simply doesn't work out.

Of course, it's up for discussion and team decision, but i would suggest to stop setting milestones on issues. One harsh way of doing this, is to remove all milestone labels at once, by simply deleting all existing milestones labels. 
Pro: It frees you from re-scheduling every single issue. You can still name and tag the next release version manually. Also, when note-worthy issues get closed, you update the changelog manually anyways. It won't impact the project in a blocking way. 
Con: The downside is that you lose a bit of guidance and signalling for the team on what to do next.

Best, Jens

**Comment by prock-fife** [comment]
> You have marked a lot of tickets belonging to the 0.4.0 milestone. When closing all the issues is a must (to be done with that milestone), you will probably never get the 0.4.0 release out.

Noted.  We have definitely experienced this first hand and appreciate your suggestions to move away from the current model to one without milestones.  I will open another issue for discussion.

> Con: The downside is that you lose a bit of guidance and signalling for the team on what to do next.

Generally the team works on what they want to work on based on a general understanding of where we want to go.  I don't think milestones have been successful in representing this anyway.

**Comment by jakoch** [comment]
I've now implemented Git commit based builds and deployments from the master branch.
The following `ReleaseArtifact`s are generated: `fifengine-dev-ca88cf0-VC.zip` and `fifengine-VC.zip`.
https://ci.appveyor.com/project/jakoch/fifengine/build/dev-ca88cf0/artifacts
This pretty much solved the issue regarding "Nightly Builds for Windows".

To progress on the MAC side, yet..

**Comment by jakoch** [comment]
Who could help doing the Mac packaging?

**Comment by vtchill** [comment]
I can try to help. I'll just need some instruction on how to get up and running with our current setup for packaging in general.

**Comment by jakoch** [comment]
Ok. 
You would build the project Travis-like. Then the last step should be "[Install the project...](https://travis-ci.org/fifengine/fifengine/jobs/171606527#L822)".
Then you could start working on an additional step for packaging. 

I could imagine that the approach from Appveyor/Windows could work on Mac, too.
We package with `python setup.py set_metadata --version=%APPVEYOR_BUILD_VERSION% bdist_wininst --target-version="%PYTHON_VERSION%"`. Of course `bdist_wininst` is wrong on Mac.
There are other distribution formats, https://docs.python.org/2/distutils/builtdist.html
But i don't know, which will work. 
When the package is created, it needs an install test.

**Comment by jakoch** [comment]
Some additional notes: https://github.com/fifengine/fifechan/issues/41
